### PR TITLE
Create scaffolding for new demand control plugin

### PR DESCRIPTION
**Pull request by tninesling**
Created scaffolding for new demand control plugin with a simple config. The max cost setting is planned to come from the limits block of the configuration, so it is not included in this change. The plugin is currently not registered so it does not appear in the config schema during development

Fixes [ROUTER-181](https://apollographql.atlassian.net/browse/ROUTER-181)

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

[ROUTER-181]: https://apollographql.atlassian.net/browse/ROUTER-181?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apollo-router/src/plugins/demand_control/mod.rs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+use tower::ServiceBuilder;
+use tower::ServiceExt;
+
+use crate::plugin::Plugin;
+use crate::plugin::PluginInit;
+use crate::services::execution::BoxService;
+
+/// Algorithm for calculating the cost of an incoming query.
+#[derive(Clone, Debug, Default, Deserialize, JsonSchema)]
+#[serde(deny_unknown_fields, rename_all = "lowercase")]
```
Comment (BrynCooke) [inline apollo-router/src/plugins/demand_control/mod.rs:14]:
```suggestion
#[serde(deny_unknown_fields, rename_all = "snake_case")]
```

File: apollo-router/src/plugins/demand_control/mod.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    /// - Subscription 0
+    ///
+    /// Type costs:
+    /// - Object: 1
+    /// - Interface: 1
+    /// - Union: 1
+    /// - Scalar: 0
+    /// - Enum: 0
+    #[default]
+    Default,
```
Comment (BrynCooke) [inline apollo-router/src/plugins/demand_control/mod.rs:30]: Lets rename this to `Basic`, and have it not defaulted. Users will have to specify the value explicitly for now.

**Review by BrynCooke — CHANGES_REQUESTED** [review]
Couple of minor things

File: apollo-router/src/plugins/demand_control/mod.rs (hunk 2 above)
Comment (tninesling) [reply apollo-router/src/plugins/demand_control/mod.rs]: https://github.com/apollographql/router/pull/4820/commits/234bff20ba6141fb2dc2fe78101d86f6c7a01036

**Review by BrynCooke — APPROVED** [review]

**Pull request merged**